File: packages/backend/src/modules/enrichment/circuit-breaker.ts
/**
 * Per-provider circuit breaker for external API resilience.
 *
 * Tracks failures in a sliding window per provider and temporarily
 * disables calls to providers that exceed the failure threshold.
 *
 * State machine:
 *   Closed → Open:     failures exceed threshold within sliding window
 *   Open → Half-Open:  cooldown period expires
 *   Half-Open → Closed: probe call succeeds
 *   Half-Open → Open:   probe call fails
 *
 * @module circuit-breaker
 */

/** Possible states of the circuit breaker for a provider. */
export type CircuitState = 'closed' | 'open' | 'half-open';

/** Injectable configuration for tuning circuit breaker behaviour. */
export interface CircuitBreakerConfig {
  /** Number of most recent calls to track per provider. */
  windowSize: number;
  /** Number of failures within the window that triggers the open state. */
  failureThreshold: number;
  /** Milliseconds to wait in open state before transitioning to half-open. */
  cooldownMs: number;
}

/** Snapshot of a single provider's circuit breaker state. */
export interface CircuitBreakerState {
  state: CircuitState;
  failureCount: number;
  lastFailureTime: number | null;
  recentCalls: Array<{ timestamp: number; success: boolean }>;
}

/** Public contract for the circuit breaker. */
export interface ICircuitBreaker {
  canCall(providerSlug: string): boolean;
  recordSuccess(providerSlug: string): void;
  recordFailure(providerSlug: string): void;
  getState(providerSlug: string): CircuitBreakerState;
  reset(providerSlug: string): void;
}

/** Default configuration values. */
const DEFAULT_CONFIG: CircuitBreakerConfig = {
  windowSize: 10,
  failureThreshold: 5,
  cooldownMs: 60_000,
};

/** Internal mutable state tracked per provider. */
interface InternalProviderState {
  state: CircuitState;
  lastFailureTime: number | null;
  recentCalls: Array<{ timestamp: number; success: boolean }>;
}

/**
 * In-memory circuit breaker that maintains independent state per provider.
 *
 * Accepts an optional config for tuning and an optional `now` function
 * (defaults to `Date.now`) so tests can control time deterministically.
 */
export class CircuitBreaker implements ICircuitBreaker {
  private readonly config: CircuitBreakerConfig;
  private readonly now: () => number;
  private readonly providers: Map<string, InternalProviderState> = new Map();

  constructor(config?: Partial<CircuitBreakerConfig>, now?: () => number) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.now = now ?? Date.now;
  }

  // ── helpers ──────────────────────────────────────────────────────────

  /** Return existing state or create a fresh closed state for the provider. */
  private getOrCreate(providerSlug: string): InternalProviderState {
    let s = this.providers.get(providerSlug);
    if (!s) {
      s = { state: 'closed', lastFailureTime: null, recentCalls: [] };
      this.providers.set(providerSlug, s);
    }
    return s;
  }

  /** Count failures in the current sliding window. */
  private failureCount(s: InternalProviderState): number {
    return s.recentCalls.filter((c) => !c.success).length;
  }

  /** Trim the sliding window to the configured size (keep most recent). */
  private trimWindow(s: InternalProviderState): void {
    if (s.recentCalls.length > this.config.windowSize) {
      s.recentCalls = s.recentCalls.slice(
        s.recentCalls.length - this.config.windowSize,
      );
    }
  }

  /** Check whether the cooldown period has elapsed for an open provider. */
  private cooldownExpired(s: InternalProviderState): boolean {
    if (s.lastFailureTime === null) return false;
    return this.now() - s.lastFailureTime >= this.config.cooldownMs;
  }

  // ── public API ───────────────────────────────────────────────────────

  /**
   * Returns `true` when the provider is available for a call.
   *
   * - Closed / half-open → true
   * - Open with expired cooldown → transitions to half-open, returns true
   * - Open with active cooldown → false
   */
  canCall(providerSlug: string): boolean {
    const s = this.getOrCreate(providerSlug);

    if (s.state === 'closed' || s.state === 'half-open') {
      return true;
    }

    // state === 'open'
    if (this.cooldownExpired(s)) {
      s.state = 'half-open';
      return true;
    }

    return false;
  }

  /**
   * Record a successful call to the provider.
   *
   * - Half-open → transitions to closed and resets the window.
   * - Closed → appends success to the sliding window.
   */
  recordSuccess(providerSlug: string): void {
    const s = this.getOrCreate(providerSlug);

    if (s.state === 'half-open') {
      // Probe succeeded — close the circuit and reset.
      s.state = 'closed';
      s.recentCalls = [];
      s.lastFailureTime = null;
      return;
    }

    s.recentCalls.push({ timestamp: this.now(), success: true });
    this.trimWindow(s);
  }

  /**
   * Record a failed call to the provider.
   *
   * - Appends failure to the sliding window.
   * - If half-open → immediately re-opens with a fresh cooldown.
   * - If closed and failures exceed threshold → opens the circuit.
   */
  recordFailure(providerSlug: string): void {
    const s = this.getOrCreate(providerSlug);
    const now = this.now();

    if (s.state === 'half-open') {
      // Probe failed — back to open with a new cooldown.
      s.state = 'open';
      s.lastFailureTime = now;
      s.recentCalls.push({ timestamp: now, success: false });
      this.trimWindow(s);
      return;
    }

    s.recentCalls.push({ timestamp: now, success: false });
    this.trimWindow(s);
    s.lastFailureTime = now;

    if (this.failureCount(s) >= this.config.failureThreshold) {
      s.state = 'open';
    }
  }

  /**
   * Return a snapshot of the provider's current circuit breaker state.
   * If the provider has never been seen, returns a fresh closed state.
   */
  getState(providerSlug: string): CircuitBreakerState {
    const s = this.getOrCreate(providerSlug);
    return {
      state: s.state,
      failureCount: this.failureCount(s),
      lastFailureTime: s.lastFailureTime,
      recentCalls: [...s.recentCalls],
    };
  }

  /**
   * Reset a provider back to the initial closed state, clearing all history.
   */
  reset(providerSlug: string): void {
    this.providers.set(providerSlug, {
      state: 'closed',
      lastFailureTime: null,
      recentCalls: [],
    });
  }
}

File: packages/scraper/src/resilience/circuit_breaker.py
"""Per-domain circuit breaker for scraping resilience.

Tracks failures per target domain using a sliding window and transitions
through closed → open → half-open states to isolate failing domains.

State machine:
- Closed → Open: failure count in sliding window exceeds threshold
- Open → Half-Open: cooldown period elapses
- Half-Open → Closed: probe request succeeds
- Half-Open → Open: probe request fails
"""

from __future__ import annotations

import time
from collections import deque
from dataclasses import dataclass, field
from enum import Enum


class CircuitState(str, Enum):
    """Circuit breaker states."""

    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class CircuitBreakerState:
    """Internal state tracked per domain."""

    domain: str
    state: CircuitState = CircuitState.CLOSED
    recent_calls: deque[tuple[float, bool]] = field(default_factory=deque)
    last_state_change: float = field(default_factory=time.monotonic)


class DomainCircuitBreaker:
    """Per-domain sliding window circuit breaker.

    Args:
        window_size: Maximum number of recent calls to track per domain.
        failure_threshold: Number of failures in the window that triggers open state.
        cooldown_seconds: Seconds to wait in open state before transitioning to half-open.
    """

    def __init__(
        self,
        window_size: int = 10,
        failure_threshold: int = 5,
        cooldown_seconds: int = 120,
    ) -> None:
        self._window_size = window_size
        self._failure_threshold = failure_threshold
        self._cooldown_seconds = cooldown_seconds
        self._states: dict[str, CircuitBreakerState] = {}

    def _get_or_create(self, domain: str) -> CircuitBreakerState:
        """Get existing state for a domain or create a new closed-state entry."""
        if domain not in self._states:
            self._states[domain] = CircuitBreakerState(domain=domain)
        return self._states[domain]

    def can_call(self, domain: str) -> bool:
        """Check whether a request to the given domain is allowed.

        - Unknown/closed domains: always allowed.
        - Open domains: allowed only if cooldown has elapsed (transitions to half-open).
        - Half-open domains: allowed (probe request).
        """
        if domain not in self._states:
            return True

        state = self._states[domain]

        if state.state == CircuitState.CLOSED:
            return True

        if state.state == CircuitState.OPEN:
            elapsed = time.monotonic() - state.last_state_change
            if elapsed >= self._cooldown_seconds:
                state.state = CircuitState.HALF_OPEN
                state.last_state_change = time.monotonic()
                return True
            return False

        # Half-open: allow the probe
        return True

    def record_success(self, domain: str) -> None:
        """Record a successful request to the domain.

        In half-open state, transitions back to closed and resets the window.
        """
        state = self._get_or_create(domain)
        now = time.monotonic()

        if state.state == CircuitState.HALF_OPEN:
            state.state = CircuitState.CLOSED
            state.last_state_change = now
            state.recent_calls.clear()
            return

        self._append_call(state, now, success=True)

    def record_failure(self, domain: str) -> None:
        """Record a failed request to the domain.

        In half-open state, transitions back to open.
        In closed state, checks if failures exceed threshold to transition to open.
        """
        state = self._get_or_create(domain)
        now = time.monotonic()

        if state.state == CircuitState.HALF_OPEN:
            state.state = CircuitState.OPEN
            state.last_state_change = now
            state.recent_calls.clear()
            return

        self._append_call(state, now, success=False)

        if state.state == CircuitState.CLOSED:
            failure_count = sum(1 for _, success in state.recent_calls if not success)
            if failure_count >= self._failure_threshold:
                state.state = CircuitState.OPEN
                state.last_state_change = now

    def get_state(self, domain: str) -> CircuitState:
        """Get the current circuit state for a domain.

        Returns CLOSED for unknown domains.
        """
        if domain not in self._states:
            return CircuitState.CLOSED
        return self._states[domain].state

    def get_all_states(self) -> dict[str, CircuitState]:
        """Get the circuit state for all tracked domains."""
        return {domain: state.state for domain, state in self._states.items()}

    def _append_call(
        self, state: CircuitBreakerState, timestamp: float, *, success: bool
    ) -> None:
        """Append a call result to the sliding window, trimming to window_size."""
        state.recent_calls.append((timestamp, success))
        while len(state.recent_calls) > self._window_size:
            state.recent_calls.popleft()
